src/Mod/ModScript.h

```diff
@@ -238,6 +239,15 @@ class ModScript
 		}
 	};
 
+	////////////////////////////////////////////////////////////
+	//					country script
+	////////////////////////////////////////////////////////////
+	// Output(fundingChange, newSatisfaction, newPact, cancelPact), Input(Country, Save, XComScore, AlienScore)
+	struct NewMonthCountryParser : ScriptParserEvents<ScriptOutputArgs<int&, int&, int&, int&>, Country*, SavedGame*, int, int>
```

Comment:
What is point to giving writable access to both game objects?
If it do not have use it should be make read-only.

Reply:
`GeoscapeGame.getRandomState` appears to require a PTRE. Likewise, all of the `SetTag` methods appear to require a PTRE. However, while the objects are non-const, neither object actually appears to offer access to any method that can mutate state outside of changing scripting values. `Country` certainly does not.

Reply:
You can pass only copy of random state. This would be beneficial as it will make global random deterministic and independent from all script logic.

And `SetTag` is this need there? as we allow `ptre` them make this function "cemented" in engine. This mean in long run any change to country or related data will be alterable in this place and engine can't reason what state is of `Country`.

This mean you could make country sign pact but script will return that country do not signed any thing, or opposite.

To sum up, what is exactly "end game" for this script? if only to return 4 values, then it should not alter anything else.

Reply:
> > You can pass only copy of random state. This would be beneficial as it will make global random deterministic and independent from all script logic.
> 
> Is this really beneficial? From the player's point of view, scripting _is_ a part of the game's state, so it should follow the same rules as the rest of the game. If you really want some sort of scriptable rng that doesn't influence the game's state, I would add something that does that to the list of "global" math functions. I'm open to writing that, though I don't see it as beneficial as opposed to following the same rules as everything else.
> 

This have noting with player. This have every thing with engine and debugging and reproducibility. If you use only once global rng all logic after will change its behavior as different value it get for rng.
If you do not touch global rng but use copy, then this script will no way affect other logic in game aside for return values.
This mean script will be near PURE functions.

> > And SetTag is this need there?
> > Well, frankly, my preference would be that script tags _always_ be mutable from a scripting context. But I haven't looked into how hard a change that would be at this point (mark all the scriptvalues mutable? I dunno). If a script can't carry forward some sort of information from invocation to invocation, it really restricts what you can do with them.
> 

As some time ago I said, script _always_ should get `const` objects aside case where whole point of script is change global state or specific object in arbitrary way.

> As for this specific hooks requirement, it's necessary to implement any logic that tracks state outside what `Country` currently enables. For example, I had the idea that a country could "leave" the X-com project due to poor performance but not sign an alien pact. Or perhaps not be in the project, to begin with, but joining the project later. This could be implemented by having an "allied" tag, which is set or unset via some internal scripting logic as appropriate. The possibilities grow as `Country` is exposed in more places, but really it's about keeping options unfettered.
> 

Question if this function should be responsible for updating global state. May be correct solution would be callback that is call at end of day for country? and in current hook you add it only do final decision.

Your current approach is problematic as can't make easy iteration between countries, you have only one chance to make decision and you could not have access to other countries or other country already make decision.

Effective to have best control over this you would need have multiple pass of script hook to correctly handle cross references.

> > This mean in long run any change to country or related data will be alterable in this place and engine can't reason what state is of Country.
> > This mean you could make country sign pact but script will return that country do not signed any thing, or opposite.
> 
> Well, I mean the logic as is can already alter the state of `Country` in more or less arbitrary ways. It can already override forming a pact or not forming a pact, change funding to an arbitrary value, changing satisfaction arbitrarily, ect. While no exposed method mutates state, the script as a whole is already mutable in more or less every way that matters.
> 

but state is change by code after this hook, and this is critical difference. if all logic would be in this script hook, then you would be 100% correct, now we have split responsibility that I would like to avoid.

> And I mean, that was my design intent that the scripting logic be able to write its own rules, so to speak, about `Country` state changes at month end. I initially wrote it in a mutable way because of this but changed it to work in a more "pure" function-like way because some minor parts of the logic were cleaner like this. But it's really a difference in style, not functionality.
> 

Yes, goal is to have arbitrary logic there, but it need have we need have bounds on what is possible (no some thing like `delete country;`) and code after `damageUnit` need run cleanups of game state of all possible changes after script.

> I think, bigger picture, it's probably more productive to handle immutability in a more "object oriented" way. Mutating state being controlled more by access or lack-there-of to methods that can mutate private data. In fact, with scripting, we can do it one better by perhaps _only_ registering "mutable" methods to the specific functions that require them (in their Parser constructor).
>
> But failing that, yeah, I think script tags need to be always writable from the context of a script. That was the main motivating factor in using a PTRE here.

No, again, they should be const as this make all code around them simple and easy to reason ablaut.
We play there long run game, scripts should not be able to corrupt game state, and this need heavy use of `const`.
I simply will not allow situation that to correctly debug game crash you need debug scripts. Using `ptre` only is asking
for more place where script could make holes in logic.

Using `ptre` make too engine code harder to refactor as we need consider all visible side effects that script can relay on.
If we add `ptre` in your new script then we could never change order in with script are called or when are called.